File: csharp/need-for-speed/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    private int _speed;
    private int _battery;
    private int _batteryDrain;
    private int _distance;

    public RemoteControlCar(int speed, int batteryDrain)
    {
        this._speed = speed;
        this._batteryDrain = batteryDrain;
    }

    public bool BatteryDrained() => _battery < _batteryDrain;

    public int DistanceDriven()
    {
        return BatteryDrained() ? _distance : _distance + _speed;
    }

    public void Drive()
    {
        if (!BatteryDrained())
        {
            _battery -= _batteryDrain;
            _distance += _speed;
        }
    }

    public static RemoteControlCar Nitro()
    {
        var car = new RemoteControlCar();
        this._speed = 50;
        this._batteryDrain = 4;
        return car;
    }
}

class RaceTrack
{
    private int _distance;

    public RaceTrack(int dst)
    {
        this._distance = dst;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!BatteryDrained())
        {
            this._distance += this._speed;
            this._battery -= this._batteryDrain;
        }

        return this._distance >= RaceTrack();
    }
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {
  private final int speed;
  private final int batteryDrain;
  private int battery;
  private int distance;

  NeedForSpeed(int speed, int batteryDrain) {
    this.speed = speed;
    this.batteryDrain = batteryDrain;
    this.battery = 100;
    this.distance = 0;
  }

  public boolean batteryDrained() {
    return this.battery < this.batteryDrain;
  }

  public int distanceDriven() {
    return this.distance;
  }

  public void drive() {
    if (!batteryDrained()) {
      this.distance += this.speed;
      this.battery -= this.batteryDrain;
    }
  }

  public static NeedForSpeed nitro() {
    return new NeedForSpeed(50, 4);
  }
}

class RaceTrack {
  private int distance;

  RaceTrack(int distance) {
    this.distance = distance;
  }

  public boolean canFinishRace(NeedForSpeed car) {
    while (car.distanceDriven() < this.distance && !car.batteryDrained()) {
      car.drive();
    }

    return car.distanceDriven() >= this.distance;
  }
}
